File: internal/parser/parse.go
package parser

import (
	"fmt"
	"strings"

	"github.com/antlr4-go/antlr/v4"
)

// ParseResult holds the components extracted from parsing a file sequence
type ParseResult struct {
	// Full path components
	Directory  string
	Basename   string
	FrameRange string
	Padding    string
	Extension  string

	// Parse type flags
	IsSequence   bool
	IsSingleFile bool
	IsPlainFile  bool
}

// errorListener collects parse errors
type errorListener struct {
	*antlr.DefaultErrorListener
	errors []error
}

func (l *errorListener) SyntaxError(recognizer antlr.Recognizer, offendingSymbol interface{},
	line, column int, msg string, e antlr.RecognitionException) {

	l.errors = append(l.errors, fmt.Errorf("parse error at %d:%d: %s", line, column, msg))
}

// ParseFileSequence parses a file sequence string and returns its components.
// Uses SLL prediction mode for better performance (30x faster), falling back to LL if needed.
func ParseFileSequence(input string) (ParseResult, error) {
	// Create the lexer
	inputStream := antlr.NewInputStream(input)
	lexer := NewfileseqLexer(inputStream)

	// Create the token stream
	stream := antlr.NewCommonTokenStream(lexer, antlr.TokenDefaultChannel)

	// Create the parser
	parser := NewfileseqParser(stream)

	// Add custom error listener to both lexer and parser
	errListener := &errorListener{}
	lexer.RemoveErrorListeners()
	lexer.AddErrorListener(errListener)
	parser.RemoveErrorListeners()
	parser.AddErrorListener(errListener)

	// Try SLL prediction mode first (much faster for unambiguous grammars)
	parser.GetInterpreter().SetPredictionMode(antlr.PredictionModeSLL)
	tree := parser.Input()

	// If SLL failed, retry with LL prediction mode (slower but more thorough)
	if len(errListener.errors) > 0 {
		// Reset for retry
		errListener.errors = errListener.errors[:0]
		inputStream = antlr.NewInputStream(input)
		lexer.SetInputStream(inputStream)
		stream.SetTokenSource(lexer)
		parser.SetInputStream(stream)

		parser.GetInterpreter().SetPredictionMode(antlr.PredictionModeLL)
		tree = parser.Input()

		// Check for parse errors after LL retry
		if len(errListener.errors) > 0 {
			return ParseResult{}, fmt.Errorf("failed to parse sequence %q: %w", input, errListener.errors[0])
		}
	}

	// Create and run the visitor
	visitor := &fileSeqVisitor{}
	tree.Accept(visitor)

	return visitor.result, nil
}

// fileSeqVisitor implements the visitor pattern to extract components
// This is TRIVIAL - just extracting what the grammar already parsed
type fileSeqVisitor struct {
	BasefileseqVisitor
	result ParseResult
}

// VisitInput dispatches to the correct child visitor
func (v *fileSeqVisitor) VisitInput(ctx *InputContext) interface{} {
	if seqCtx := ctx.Sequence(); seqCtx != nil {
		return v.VisitSequence(seqCtx.(*SequenceContext))
	}
	if poCtx := ctx.PatternOnly(); poCtx != nil {
		return v.VisitPatternOnly(poCtx.(*PatternOnlyContext))
	}
	if sfCtx := ctx.SingleFrame(); sfCtx != nil {
		return v.VisitSingleFrame(sfCtx.(*SingleFrameContext))
	}
	if pfCtx := ctx.PlainFile(); pfCtx != nil {
		return v.VisitPlainFile(pfCtx.(*PlainFileContext))
	}
	return nil
}

// VisitSequence - grammar already separated all components
func (v *fileSeqVisitor) VisitSequence(ctx *SequenceContext) interface{} {
	v.result.IsSequence = true

	// Extract directory (grammar already parsed it)
	if dirCtx := ctx.Directory(); dirCtx != nil {
		v.result.Directory = dirCtx.GetText()
	}

	// Extract basename (grammar already parsed it)
	if basenameCtx := ctx.SequenceBasename(); basenameCtx != nil {
		v.result.Basename = basenameCtx.GetText()
	}

	// Extract frame range (grammar already parsed it)
	if frameRangeCtx := ctx.FrameRange(); frameRangeCtx != nil {
		v.result.FrameRange = frameRangeCtx.GetText()
	}

	// Post-process: move leading dot from frame range to basename
	// (see parse_postprocess.go Special Case 3 for detailed explanation)
	moveLeadingDotFromFrameRange(&v.result)

	// Extract padding (grammar already parsed it)
	if padCtx := ctx.Padding(); padCtx != nil {
		v.result.Padding = padCtx.GetText()
	}

	// Extract extensions (grammar already parsed them)
	v.result.Extension = v.extractExtensions(ctx.AllExtension())

	return nil
}

// VisitSingleFrame - grammar already separated all components
func (v *fileSeqVisitor) VisitSingleFrame(ctx *SingleFrameContext) interface{} {
	v.result.IsSingleFile = true

	// Extract directory (grammar already parsed it)
	if dirCtx := ctx.Directory(); dirCtx != nil {
		v.result.Directory = dirCtx.GetText()
	}

	// Extract basename (grammar already parsed it)
	if basenameCtx := ctx.SingleFrameBasename(); basenameCtx != nil {
		v.result.Basename = basenameCtx.GetText()
	}

	// Extract frame number (grammar already parsed it)
	if frameNumCtx := ctx.FrameNum(); frameNumCtx != nil {
		v.result.FrameRange = frameNumCtx.GetText()
	}

	// Post-process: move leading dot from frame range to basename
	// (see parse_postprocess.go Special Case 3 for detailed explanation)
	moveLeadingDotFromFrameRange(&v.result)

	// Extract extensions (grammar already parsed them)
	v.result.Extension = v.extractExtensions(ctx.AllExtension())

	// Post-process: handle special case of multiple DOT_NUM tokens
	// (see parse_postprocess.go Special Case 3 for detailed explanation)
	handleMultipleDotNumTokens(&v.result)

	return nil
}

// VisitPatternOnly - pattern-only sequence (padding without frame range)
func (v *fileSeqVisitor) VisitPatternOnly(ctx *PatternOnlyContext) interface{} {
	v.result.IsSequence = true // Pattern-only is still a sequence type

	// Extract directory
	if dirCtx := ctx.Directory(); dirCtx != nil {
		v.result.Directory = dirCtx.GetText()
	}

	// Extract basename
	if basenameCtx := ctx.PatternBasename(); basenameCtx != nil {
		v.result.Basename = basenameCtx.GetText()
	}

	// No frame range for pattern-only (initially)

	// Extract padding
	if padCtx := ctx.Padding(); padCtx != nil {
		v.result.Padding = padCtx.GetText()
	}

	// Extract extensions
	v.result.Extension = v.extractExtensions(ctx.AllExtension())

	// Post-process: extract digits that were absorbed into extension token
	// (see parse_postprocess.go for detailed explanation)
	extractDigitsFromExtensionToken(&v.result)

	return nil
}

// VisitPlainFile - grammar already separated all components
func (v *fileSeqVisitor) VisitPlainFile(ctx *PlainFileContext) interface{} {
	v.result.IsPlainFile = true

	// Extract directory (grammar already parsed it)
	if dirCtx := ctx.Directory(); dirCtx != nil {
		v.result.Directory = dirCtx.GetText()
	}

	// Extract basename (grammar already parsed it)
	if basenameCtx := ctx.PlainBasename(); basenameCtx != nil {
		v.result.Basename = basenameCtx.GetText()
	}

	// Extract extensions (grammar already parsed them)
	v.result.Extension = v.extractExtensions(ctx.AllExtension())

	return nil
}

// extractExtensions concatenates all extension tokens
func (v *fileSeqVisitor) extractExtensions(extCtxs []IExtensionContext) string {
	if len(extCtxs) == 0 {
		return ""
	}

	var sb strings.Builder
	for _, extCtx := range extCtxs {
		sb.WriteString(extCtx.GetText())
	}
	return sb.String()
}

File: cpp/private/parser/parse.cpp
#include "parse.h"
#include "parse_postprocess.h"
#include "fileseqLexer.h"
#include "fileseqParser.h"
#include "fileseqBaseVisitor.h"

#include "antlr4-runtime.h"
#include "atn/ParserATNSimulator.h"
#include "atn/PredictionMode.h"

#include <sstream>

namespace fileseq {
namespace internal {

// Custom visitor that extracts components from the parse tree
class FileSequenceVisitor : public fileseqBaseVisitor {
public:
    ParseResult result;
    std::string error;

    // Visit sequence: dir + basename? + frameRange + padding + extension*
    virtual antlrcpp::Any visitSequence(fileseqParser::SequenceContext *ctx) override {
        result.isSequence = true;

        // Extract directory
        if (ctx->directory()) {
            result.directory = ctx->directory()->getText();
        }

        // Extract basename (optional)
        if (ctx->sequenceBasename()) {
            result.basename = ctx->sequenceBasename()->getText();
        }

        // Check for unsupported subframe syntax (Python-only feature)
        // Subframes use: file.1-5#.10-20@@.exr (dual range) or file.1-5@.#.exr (composite padding)
        if (ctx->frameRange().size() > 1 || ctx->padding().size() > 1) {
            error = "subframe syntax not supported: " + ctx->getText();
            return nullptr;
        }

        // Extract frame range
        if (ctx->frameRange(0)) {
            result.frameRange = ctx->frameRange(0)->getText();
        }

        // Extract padding
        if (ctx->padding(0)) {
            result.padding = ctx->padding(0)->getText();
        }

        // Extract extension (can be multiple parts)
        if (!ctx->extension().empty()) {
            std::ostringstream ext;
            for (auto extCtx : ctx->extension()) {
                ext << extCtx->getText();
            }
            result.extension = ext.str();
        }

        return nullptr;
    }

    // Visit patternOnly: dir + basename? + padding + extension*
    virtual antlrcpp::Any visitPatternOnly(fileseqParser::PatternOnlyContext *ctx) override {
        result.isSequence = true;

        // Extract directory
        if (ctx->directory()) {
            result.directory = ctx->directory()->getText();
        }

        // Extract basename (optional)
        if (ctx->patternBasename()) {
            result.basename = ctx->patternBasename()->getText();
        }

        // Check for unsupported subframe syntax (Python-only feature)
        // Pattern-only subframes use: file.#.@@.exr (composite padding)
        if (ctx->padding().size() > 1) {
            error = "subframe syntax not supported: " + ctx->getText();
            return nullptr;
        }

        // Extract padding
        if (ctx->padding(0)) {
            result.padding = ctx->padding(0)->getText();
        }

        // Extract extension (can be multiple parts)
        if (!ctx->extension().empty()) {
            std::ostringstream ext;
            for (auto extCtx : ctx->extension()) {
                ext << extCtx->getText();
            }
            result.extension = ext.str();
        }

        return nullptr;
    }

    // Visit singleFrame: dir + basename + frameNum + extension*
    virtual antlrcpp::Any visitSingleFrame(fileseqParser::SingleFrameContext *ctx) override {
        result.isSingleFile = true;

        // Extract directory
        if (ctx->directory()) {
            result.directory = ctx->directory()->getText();
        }

        // Extract basename
        if (ctx->singleFrameBasename()) {
            result.basename = ctx->singleFrameBasename()->getText();
        }

        // Extract frame number
        if (ctx->frameNum()) {
            result.frameRange = ctx->frameNum()->getText();
        }

        // Extract extension (can be multiple parts)
        if (!ctx->extension().empty()) {
            std::ostringstream ext;
            for (auto extCtx : ctx->extension()) {
                ext << extCtx->getText();
            }
            result.extension = ext.str();
        }

        return nullptr;
    }

    // Visit plainFile: dir + basename? + extension*
    virtual antlrcpp::Any visitPlainFile(fileseqParser::PlainFileContext *ctx) override {
        result.isPlainFile = true;

        // Extract directory
        if (ctx->directory()) {
            result.directory = ctx->directory()->getText();
        }

        // Extract basename (optional)
        if (ctx->plainBasename()) {
            result.basename = ctx->plainBasename()->getText();
        }

        // Extract extension (can be multiple parts)
        if (!ctx->extension().empty()) {
            std::ostringstream ext;
            for (auto extCtx : ctx->extension()) {
                ext << extCtx->getText();
            }
            result.extension = ext.str();
        }

        return nullptr;
    }
};


bool parseFileSequence(ParseResult& result, const std::string& path) {
    try {
        // Create input stream from string
        antlr4::ANTLRInputStream input(path);

        // Create lexer
        fileseqLexer lexer(&input);

        // Remove default error listeners from lexer
        lexer.removeErrorListeners();

        // Create token stream
        antlr4::CommonTokenStream tokens(&lexer);

        // Create parser
        fileseqParser parser(&tokens);

        // Remove default error listeners from parser
        parser.removeErrorListeners();

        // Try SLL prediction mode first (30x faster for unambiguous grammars)
        parser.getInterpreter<antlr4::atn::ParserATNSimulator>()->setPredictionMode(
            antlr4::atn::PredictionMode::SLL);

        // Parse the input
        fileseqParser::InputContext* tree = parser.input();

        // If SLL failed, retry with LL prediction mode (slower but more thorough)
        if (lexer.getNumberOfSyntaxErrors() > 0 || parser.getNumberOfSyntaxErrors() > 0) {
            // Reset for retry
            input.seek(0);
            lexer.setInputStream(&input);
            tokens.setTokenSource(&lexer);
            parser.setInputStream(&tokens);

            parser.getInterpreter<antlr4::atn::ParserATNSimulator>()->setPredictionMode(
                antlr4::atn::PredictionMode::LL);

            tree = parser.input();

            // Check for parse errors after LL retry
            if (lexer.getNumberOfSyntaxErrors() > 0 || parser.getNumberOfSyntaxErrors() > 0) {
                return false;
            }
        }

        // Create visitor and walk the tree
        FileSequenceVisitor visitor;
        visitor.visit(tree);

        // Check for visitor errors (e.g., unsupported subframe syntax)
        if (!visitor.error.empty()) {
            return false;
        }

        // Get the result
        result = visitor.result;

        // Apply post-processing for the 3 special cases
        postProcessParseResult(result);

        return true;

    } catch (...) {
        // Any exception during parsing is a failure
        return false;
    }
}

} // namespace internal
} // namespace fileseq
